feat: Limit `ExpireTrash` job to 30 minutes

artonge opened the pull request:
1. feat: Implements `getSeenUsers` to iterate over users
2. chore: Refactor `callForSeenUsers` to use `getSeenUsers`
3. feat: Limit `ExpireTrash` job to 30 minutes

susnux commented on lib/private/User/Manager.php:
```diff
@@ -628,30 +628,35 @@ public function countSeenUsers() {
 					}
 				}
+
+				$offset++;
```
I do not think this is reliable, no?
Imagine you add a user in between / a user gets seen in between.

Also ordering in `getSeenUserIds` will not work if a user logs in between attempts.

@ChristophWurst posted a really interesting approach lately about how to do proper chunking on the database.
I think we should have a `getUsersAfter(timestamp)` method that gets all users with a `firstLoginTimestamp` after the specified `timestamp` parameter.

This will also work if a user logs-in in between two attempts to `callForSeen`

susnux replied:
(`lastSeen` instead of `firstSeen` also works but then a user could be returned multiple times because the timestamp would update).

susnux replied:
>  The issue already exists if a user is added while a job is running, it will only be picked up by the next job.

The bigger problem here is that the order is not the same, as currently there is no ordering the database might return the values in random order so with this PR even if no user logs in or is added some users might be lost because the database might return them in different order (no `ORDER BY` in the db query).

susnux replied:
(BTW `getLastLoggedInUsers` is not an alternative as this uses again simple offset instead of the timestamp to compare, so the results will miss users)

artonge replied:
As agreed, I'll keep the current implementation of `callForSeenUsers`. This is the best way for now to target all users.
- Order is not specified, but should be consistent across calls.
- Targeting `firstLogin` is not possible because it was introduces recently and is `-1` for most user.
- Targeting `lastLogin` exposes some user to not being picked up as if a user logs in before the run is over, then all users are shifted back, and the offset will miss some of them.

come-nc commented on lib/public/IUserManager.php:
```diff
@@ -200,9 +200,12 @@ public function countSeenUsers();
 	 * @psalm-param \Closure(\OCP\IUser):?bool $callback
+	 * @param int|null $maxTime - Returns early if time() is greater than this value.
+	 * @param int $offset - The offset from which to start.
+	 * @return int - The offset from which to continue if we returned early.
```
```suggestion
	 * @return int - The offset from which to continue if we returned early, 0 otherwise.
```

artonge replied:
Any added benefit added from using DateTime?

come-nc reviewed (approved):
Just wondering if we should use DateTime instead of int :shrug:

icewind1991 commented:
Instead of adding a limit to `callForSeensUsers` it might be better to instead have a method that returns all (seen) users as an iterator. Then the call site can do it's own limiting.

artonge commented:
> Instead of adding a limit to `callForSeensUsers` it might be better to instead have a method that returns all (seen) users as an iterator. Then the call site can do it's own limiting.

Good idea and easy to implement :). Just pushed that version in the latest commit. Will remove the other change after review.

Note that this probably need to be backported

icewind1991 commented:
Looks good, you can also re-implement `callForSeenUsers` on top of `getAllSeenUsers` for some de-duplication

artonge committed:
chore: Refactor callForSeenUsers to use getSeenUsers

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge committed:
feat: Limit trash expire job to 30 minutes

And pick up where it left off, leveraging getSeenUsers.

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge committed:
feat: Implement getSeenUsers to iterate over users

This method uses an iterator.
This is lighter on resources and gives more control to the caller

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented:
/backport to stable30

artonge commented:
/backport to stable31

artonge commented:
/backport to stable29
